File: FR-BAR-RT-main/src/services/reportService.ts
import axios from 'axios'

const API_BASE_URL = '/api/reports'

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
})

// Interceptor para agregar token automáticamente
api.interceptors.request.use((config) => {
  const token = localStorage.getItem('token')
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
})

export interface UserReport {
  total_users: number
  active_users: number
  inactive_users: number
  users_by_role: Array<{
    role_code: string
    role_name: string
    count: number
  }>
  users_by_location: Array<{
    location_id: string
    location_name: string
    count: number
  }>
  recent_logins: number
  failed_logins: number
}

export interface ProductReport {
  total_products: number
  active_products: number
  inactive_products: number
  products_by_category: Array<{
    category_id: string
    category_name: string
    count: number
  }>
  average_price: number
  price_range: {
    min: number
    max: number
  }
}

export interface VenueReport {
  total_locations: number
  active_locations: number
  total_tables: number
  tables_by_status: Array<{
    status: string
    count: number
  }>
  location_details: Array<{
    location_id: string
    location_name: string
    code: string
    total_tables: number
    available_tables: number
    occupied_tables: number
    reserved_tables: number
  }>
}

export interface ActivityReport {
  total_login_attempts: number
  successful_logins: number
  failed_logins: number
  login_attempts_by_day: Array<{
    date: string
    count: number
  }>
  recent_failed_logins: Array<{
    email: string
    ip_address: string
    attempted_at: string
  }>
  password_resets: number
  password_resets_by_day: Array<{
    date: string
    count: number
  }>
}

export interface VenueSpecificReport {
  location_id: string
  location_name: string
  location_code: string
  total_tables: number
  tables_status: {
    status: string
    count: number
  }
  assigned_users: number
  user_details: Array<{
    user_id: string
    email: string
    first_name: string
    last_name: string
    roles: string[]
    is_active: boolean
  }>
}

export interface AnalyticsReport {
  generated_at: string
  user_report: UserReport
  product_report: ProductReport
  venue_report: VenueReport
  activity_report: ActivityReport
}

export const reportService = {
  async getAnalyticsReport(): Promise<AnalyticsReport> {
    const response = await api.get('/analytics')
    return response.data
  },

  async getUserReport(): Promise<UserReport> {
    const response = await api.get('/users')
    return response.data
  },

  async getProductReport(): Promise<ProductReport> {
    const response = await api.get('/products')
    return response.data
  },

  async getVenueReport(): Promise<VenueReport> {
    const response = await api.get('/venues')
    return response.data
  },

  async getActivityReport(days?: number): Promise<ActivityReport> {
    const params = days ? { days } : {}
    const response = await api.get('/activity', { params })
    return response.data
  },

  async getVenueSpecificReport(locationId: string): Promise<VenueSpecificReport> {
    const response = await api.get(`/venue/${locationId}`)
    return response.data
  },
}

File: MS-REPORTS-GO-main/internal/service/report_service.go
package service

import (
	"ms-reports-go/internal/models"
	"ms-reports-go/internal/repository"
	"time"
)

type ReportService interface {
	GetUserReport() (*models.UserReport, error)
	GetProductReport() (*models.ProductReport, error)
	GetVenueReport() (*models.VenueReport, error)
	GetActivityReport(days int) (*models.ActivityReport, error)
	GetVenueSpecificReport(locationID string) (*models.VenueSpecificReport, error)
	GetAnalyticsReport() (*models.AnalyticsReport, error)
}

type reportService struct {
	repo repository.ReportRepository
}

func NewReportService(repo repository.ReportRepository) ReportService {
	return &reportService{repo: repo}
}

func (s *reportService) GetUserReport() (*models.UserReport, error) {
	totalUsers, err := s.repo.GetTotalUsers()
	if err != nil {
		return nil, err
	}

	activeUsers, err := s.repo.GetActiveUsers()
	if err != nil {
		return nil, err
	}

	usersByRole, err := s.repo.GetUsersByRole()
	if err != nil {
		return nil, err
	}

	usersByLocation, err := s.repo.GetUsersByLocation()
	if err != nil {
		return nil, err
	}

	recentLogins, err := s.repo.GetRecentLogins(7)
	if err != nil {
		return nil, err
	}

	failedLogins, err := s.repo.GetFailedLogins(7)
	if err != nil {
		return nil, err
	}

	return &models.UserReport{
		TotalUsers:      totalUsers,
		ActiveUsers:     activeUsers,
		InactiveUsers:   totalUsers - activeUsers,
		UsersByRole:     usersByRole,
		UsersByLocation: usersByLocation,
		RecentLogins:    recentLogins,
		FailedLogins:    failedLogins,
	}, nil
}

func (s *reportService) GetProductReport() (*models.ProductReport, error) {
	totalProducts, err := s.repo.GetTotalProducts()
	if err != nil {
		return nil, err
	}

	activeProducts, err := s.repo.GetActiveProducts()
	if err != nil {
		return nil, err
	}

	productsByCategory, err := s.repo.GetProductsByCategory()
	if err != nil {
		return nil, err
	}

	averagePrice, err := s.repo.GetAveragePrice()
	if err != nil {
		return nil, err
	}

	priceRange, err := s.repo.GetPriceRange()
	if err != nil {
		return nil, err
	}

	return &models.ProductReport{
		TotalProducts:      totalProducts,
		ActiveProducts:     activeProducts,
		InactiveProducts:   totalProducts - activeProducts,
		ProductsByCategory: productsByCategory,
		AveragePrice:       averagePrice,
		PriceRange:         priceRange,
	}, nil
}

func (s *reportService) GetVenueReport() (*models.VenueReport, error) {
	totalLocations, err := s.repo.GetTotalLocations()
	if err != nil {
		return nil, err
	}

	activeLocations, err := s.repo.GetActiveLocations()
	if err != nil {
		return nil, err
	}

	totalTables, err := s.repo.GetTotalTables()
	if err != nil {
		return nil, err
	}

	tablesByStatus, err := s.repo.GetTablesByStatus()
	if err != nil {
		return nil, err
	}

	locationDetails, err := s.repo.GetLocationDetails()
	if err != nil {
		return nil, err
	}

	return &models.VenueReport{
		TotalLocations:  totalLocations,
		ActiveLocations: activeLocations,
		TotalTables:     totalTables,
		TablesByStatus:  tablesByStatus,
		LocationDetails: locationDetails,
	}, nil
}

func (s *reportService) GetActivityReport(days int) (*models.ActivityReport, error) {
	if days <= 0 {
		days = 30 // Default to 30 days
	}

	totalLoginAttempts, err := s.repo.GetTotalLoginAttempts(days)
	if err != nil {
		return nil, err
	}

	successfulLogins, err := s.repo.GetSuccessfulLogins(days)
	if err != nil {
		return nil, err
	}

	failedLogins, err := s.repo.GetFailedLogins(days)
	if err != nil {
		return nil, err
	}

	loginAttemptsByDay, err := s.repo.GetLoginAttemptsByDay(days)
	if err != nil {
		return nil, err
	}

	recentFailedLogins, err := s.repo.GetRecentFailedLogins(10)
	if err != nil {
		return nil, err
	}

	passwordResets, err := s.repo.GetPasswordResets(days)
	if err != nil {
		return nil, err
	}

	passwordResetsByDay, err := s.repo.GetPasswordResetsByDay(days)
	if err != nil {
		return nil, err
	}

	return &models.ActivityReport{
		TotalLoginAttempts:  totalLoginAttempts,
		SuccessfulLogins:    successfulLogins,
		FailedLogins:        failedLogins,
		LoginAttemptsByDay:  loginAttemptsByDay,
		RecentFailedLogins:  recentFailedLogins,
		PasswordResets:      passwordResets,
		PasswordResetsByDay: passwordResetsByDay,
	}, nil
}

func (s *reportService) GetVenueSpecificReport(locationID string) (*models.VenueSpecificReport, error) {
	return s.repo.GetVenueDetails(locationID)
}

func (s *reportService) GetAnalyticsReport() (*models.AnalyticsReport, error) {
	userReport, err := s.GetUserReport()
	if err != nil {
		return nil, err
	}

	productReport, err := s.GetProductReport()
	if err != nil {
		return nil, err
	}

	venueReport, err := s.GetVenueReport()
	if err != nil {
		return nil, err
	}

	activityReport, err := s.GetActivityReport(30)
	if err != nil {
		return nil, err
	}

	return &models.AnalyticsReport{
		GeneratedAt:    time.Now(),
		UserReport:     *userReport,
		ProductReport:  *productReport,
		VenueReport:    *venueReport,
		ActivityReport: *activityReport,
	}, nil
}




